### PR TITLE
[OCPQE-12211] Enable WINC QE CI workflow for AWS IPI.

### PR DESCRIPTION
**Pull request by jfrancoa**

### CODE REVIEW

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-master.yaml [hunk 1]
```diff
@@ -166,6 +166,28 @@ tests:
     test:
     - ref: openshift-extended-test
     workflow: cucushift-installer-rehearse-aws-ipi-private
+- as: debug-winc-aws-ipi
+  optional: true
+  run_if_changed: test/extended/winc/
+  steps:
+    cluster_profile: aws-qe
+    dependency_overrides:
+      BUNDLE_IMAGE: quay.io/winc/wmco-bundle:latest
```
Comment (jianzhangbjz) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-master.yaml:175]: Where this bundle image from? I didn't see any step for this bundle image building.

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-master.yaml [hunk 2]
```diff
@@ -166,6 +166,28 @@ tests:
...
+      OO_CHANNEL: stable
+      OO_INSTALL_NAMESPACE: openshift-windows-machine-config-operator
+      OO_PACKAGE: windows-machine-config-operator
+      OO_TARGET_NAMESPACES: '!install'
+      SUB_INSTALL_NAMESPACE: openshift-windows-machine-config-operator
+      TEST_FILTERS: ~ChkUpgrade&;~ConnectedOnly&
+      TEST_SCENARIOS: Windows_Containers
+      TEST_TIMEOUT: "50"
+    test:
+    - ref: openshift-extended-test
```
Comment (jianzhangbjz) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-master.yaml:189]: If you want to run this e2e, you don't need to add the above steps. For QE automation, as you know, your operator bundles will be stored in the quay.io/openshift-qe-optional-operators/ocp4-index:latest image. So, we enabled this step: https://github.com/openshift/release/tree/master/ci-operator/step-registry/enable-qe-catalogsource for the OLM operators sub team. So, for QE, you only need to contribute your e2e tests in the https://github.com/openshift/openshift-tests-private, we already integrated it to the Prow CI.

**Review by jianzhangbjz — CHANGES_REQUESTED** [review]

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-master.yaml (hunk 1 above)
Comment (jianzhangbjz) [reply ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-master.yaml:175]: If you want to run those e2e cases of https://github.com/openshift/openshift-tests-private, you don’t need to do that.

**Comment by jfrancoa** [comment]
/pj-rehearse pull-ci-openshift-openshift-tests-private-release-4.12-debug-winc-aws-ipi

**Comment by jfrancoa** [comment]
/pj-rehearse pull-ci-openshift-openshift-tests-private-release-4.12-debug-winc-aws-ipi

**Comment by jfrancoa** [comment]
/pj-rehearse pull-ci-openshift-openshift-tests-private-release-4.12-debug-winc-aws-ipi

File: ci-operator/step-registry/openshift-qe/winc/openshift-qe-winc-commands.sh [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+winworker_machineset_name=$(oc get machineset -n openshift-machine-api -o json | jq -r '.items[] | select(.metadata.name | test("win")).metadata.name')
+winworker_machineset_replicas=$(oc get machineset -n openshift-machine-api $winworker_machineset_name -o jsonpath="{.spec.replicas}")
+
+echo "Waiting for Windows nodes to come up"
+while [[ $(oc -n openshift-machine-api get machineset/${winworker_machineset_name} -o 'jsonpath={.status.readyReplicas}') != "${winworker_machineset_replicas}" ]]; do echo -n "." && sleep 10; done
+#oc wait machineset $winworker_machineset_name -n openshift-machine-api --for jsonpath="{.status.fullyLabeledReplicas}"=$winworker_machineset_replicas --timeout=5m
+#oc wait machine -l machine.openshift.io/cluster-api-machineset=${winworker_machineset_name} -n openshift-machine-api --for jsonpath="{.status.phase}"=Running --timeout=30m
+
+oc wait nodes -l kubernetes.io/os=windows --for condition=Ready=True --timeout=15m
```
Comment (rrasouli) [inline ci-operator/step-registry/openshift-qe/winc/openshift-qe-winc-commands.sh:137]: Are you sure that 15 minutes is enough to create Windows workers? or it is the limit in Prow?

File: ci-operator/step-registry/openshift-qe/winc/openshift-qe-winc-commands.sh [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+IAAS_PLATFORM=$(oc get infrastructure cluster -o=jsonpath="{.status.platformStatus.type}"| tr '[:upper:]' '[:lower:]')
+
+
+winworker_machineset_name=$(oc get machineset -n openshift-machine-api -o json | jq -r '.items[] | select(.metadata.name | test("win")).metadata.name')
+winworker_machineset_replicas=$(oc get machineset -n openshift-machine-api $winworker_machineset_name -o jsonpath="{.spec.replicas}")
+
+echo "Waiting for Windows nodes to come up"
+while [[ $(oc -n openshift-machine-api get machineset/${winworker_machineset_name} -o 'jsonpath={.status.readyReplicas}') != "${winworker_machineset_replicas}" ]]; do echo -n "." && sleep 10; done
+#oc wait machineset $winworker_machineset_name -n openshift-machine-api --for jsonpath="{.status.fullyLabeledReplicas}"=$winworker_machineset_replicas --timeout=5m
+#oc wait machine -l machine.openshift.io/cluster-api-machineset=${winworker_machineset_name} -n openshift-machine-api --for jsonpath="{.status.phase}"=Running --timeout=30m
```
[the remark this answers is not in the document]
Comment (rrasouli) [reply ci-operator/step-registry/openshift-qe/winc/openshift-qe-winc-commands.sh]: No need for more than 10 minutes waiting for nodes to be ready then

**Comment by jfrancoa** [comment]
@jrvaldes Could you please give your feedback? mostly on the changes done in ci-operator/step-registry/ipi/conf/aws/windows-machineset/ipi-conf-aws-windows-machineset-commands.sh and ci-operator/step-registry/openshift-qe/winc/openshift-qe-winc-commands.sh ? 
I observed that in some cases the windows-user-data check was failing, mostly because WMCO was not up and running yet. That is why I added the wait for deployment windows-machine-config-operator/openshift-windows-machine-config-operator. And regarding the OS image, probably soon we will be supporting Windows 2022 for AWS too..so I though it would be interesting parametrizing that, so that we can have several jobs with different Windows versions.

File: ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.10.yaml [hunk 5]
```diff
@@ -11,6 +11,10 @@ base_images:
     name: "4.10"
     namespace: ocp
     tag: tools
+  upi-installer:
```
Comment (jfrancoa) [inline ci-operator/config/openshift/openshift-tests-private/openshift-openshift-tests-private-release-4.10.yaml:14]: I added upi-installer as it is needed by the ipi-conf-aws-windows-machineset ref and it is included on the other config files for 4.11, 4.12, 4.13 and master.

**Comment by jfrancoa** [comment]
/pj-rehearse pull-ci-openshift-openshift-tests-private-release-4.13-debug-winc-aws-ipi

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/winc/deprovision/cucushift-installer-rehearse-aws-ipi-winc-deprovision-chain.metadata.json [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+		],
+		"reviewers": [
+			"jianlinliu",
+			"yunjiang29",
+			"gpei",
+			"jfrancoa",
+			"rrasouli"
+		]
+	}
+}
```
Comment (rrasouli) [inline ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/winc/deprovision/cucushift-installer-rehearse-aws-ipi-winc-deprovision-chain.metadata.json:17]: New line

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/winc/provision/cucushift-installer-rehearse-aws-ipi-winc-provision-chain.metadata.json [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+		],
+		"reviewers": [
+			"jianlinliu",
+			"yunjiang29",
+			"gpei",
+			"jfrancoa",
+			"rrasouli"
+		]
+	}
+}
```
Comment (rrasouli) [inline ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/winc/provision/cucushift-installer-rehearse-aws-ipi-winc-provision-chain.metadata.json:17]: ditto

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/winc/deprovision/cucushift-installer-rehearse-aws-ipi-winc-deprovision-chain.metadata.json (hunk 6 above)
Comment (jfrancoa) [reply ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/winc/deprovision/cucushift-installer-rehearse-aws-ipi-winc-deprovision-chain.metadata.json:17]: You can check other metadata files from the repo, like https://github.com/openshift/release/blob/master/ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/kubevirt/provision/cucushift-installer-rehearse-aws-ipi-ovn-kubevirt-provision-chain.metadata.json, for example, and none of them have extra new line.

**Comment by kuiwang02** [comment]
i am ok from golang's view.

/lgtm

**Comment by jfrancoa** [comment]
pinging @jianlinliu  for approval

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/winc/cucushift-installer-rehearse-aws-ipi-winc-workflow.yaml [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+      SUB_PACKAGE: windows-machine-config-operator
+      SUB_TARGET_NAMESPACES: '!install'
+      SUB_SOURCE: "qe-app-registry"
+      WINDOWS_NODE_REPLICAS: 2
+    pre:
+    - chain: cucushift-installer-rehearse-aws-ipi-winc-provision
+    test:
+    - ref: cucushift-installer-check
+    post:
+    - chain: cucushift-installer-rehearse-aws-ipi-winc-deprovision
```
Comment (LiangquanLi930) [inline ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/winc/cucushift-installer-rehearse-aws-ipi-winc-workflow.yaml:16]: Should use `cucushift-installer-rehearse-aws-ipi-deprovision`, If there is no special operation, we don't need to define additional chain

Comment (jfrancoa) [reply ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/winc/cucushift-installer-rehearse-aws-ipi-winc-workflow.yaml:16]: @LiangquanLi930 So, that means I can get rid of the whole deprovision folder right?

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/winc/provision/cucushift-installer-rehearse-aws-ipi-winc-provision-chain.yaml [hunk 9]
```diff
@@ -0,0 +1,15 @@
+chain:
+  as: cucushift-installer-rehearse-aws-ipi-winc-provision
+  steps:
+  - chain: ipi-conf-aws
```
Comment (LiangquanLi930) [inline ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/winc/provision/cucushift-installer-rehearse-aws-ipi-winc-provision-chain.yaml:4]: We should change to 
```
- chain: cucushift-installer-rehearse-aws-ipi-ovn-provision
- ref: ovn-conf-hybrid-manifest
- ref: ssh-bastion
......
```

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/winc/provision/cucushift-installer-rehearse-aws-ipi-winc-provision-chain.metadata.json [hunk 10]
```diff
@@ -0,0 +1,17 @@
+{
+	"path": "cucushift/installer/rehearse/aws/ipi/winc/provision/cucushift-installer-rehearse-aws-ipi-winc-provision-chain.yaml",
```
Comment (LiangquanLi930) [inline ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/winc/provision/cucushift-installer-rehearse-aws-ipi-winc-provision-chain.metadata.json:2]: this ocp is ovn, so we need mv to `ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/winc`

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/winc/provision/cucushift-installer-rehearse-aws-ipi-winc-provision-chain.yaml (hunk 9 above)
Comment (jfrancoa) [reply ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/winc/provision/cucushift-installer-rehearse-aws-ipi-winc-provision-chain.yaml:4]: The[ step adds a yaml which is then used during installation](https://github.com/openshift/release/blob/master/ci-operator/step-registry/ovn/conf/hybrid-manifest/ovn-conf-hybrid-manifest-commands.sh), if we run the step after ipi-install we won't have enabled OVN Hybrid overlay

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/winc/provision/cucushift-installer-rehearse-aws-ipi-winc-provision-chain.metadata.json (hunk 10 above)
Comment (LiangquanLi930) [reply ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/winc/provision/cucushift-installer-rehearse-aws-ipi-winc-provision-chain.metadata.json:2]: yes

**Comment by jfrancoa** [comment]
/pj-rehearse pull-ci-openshift-openshift-tests-private-release-4.12-debug-winc-aws-ipi

**Comment by jfrancoa** [comment]
/pj-rehearse pull-ci-openshift-openshift-tests-private-release-4.12-debug-winc-aws-ipi

**Comment by jfrancoa** [comment]
/pj-rehearse pull-ci-openshift-openshift-tests-private-master-debug-winc-aws-ipi

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/winc/provision/cucushift-installer-rehearse-aws-ipi-ovn-winc-provision-chain.yaml [hunk 11]
```diff
@@ -0,0 +1,15 @@
...
+  - chain: ipi-conf-aws
+  - ref: ovn-conf
+  - ref: ovn-conf-hybrid-manifest
+  - chain: ipi-install
+  - ref: enable-qe-catalogsource
+  - ref: ssh-bastion
+  - ref: operatorhub-subscribe
+  - ref: windows-conf-operator
+  - ref: ipi-conf-aws-windows-machineset
+  - ref: openshift-qe-winc
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/winc/provision/cucushift-installer-rehearse-aws-ipi-ovn-winc-provision-chain.yaml:13]: Sounds like the `openshift-qe-winc` is for test purpose? Is it a MUST for the whole install workflow?

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/winc/cucushift-installer-rehearse-aws-ipi-ovn-winc-workflow.yaml [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+      SUB_PACKAGE: windows-machine-config-operator
+      SUB_TARGET_NAMESPACES: '!install'
+      SUB_SOURCE: "qe-app-registry"
+      WINDOWS_NODE_REPLICAS: 2
+    pre:
+    - chain: cucushift-installer-rehearse-aws-ipi-ovn-winc-provision
+    test:
+    - ref: cucushift-installer-check
+    post:
+    - chain: cucushift-installer-rehearse-aws-ipi-deprovision
```
Comment (jianlinliu) [inline ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/winc/cucushift-installer-rehearse-aws-ipi-ovn-winc-workflow.yaml:16]: I am wondering the `cucushift-installer-rehearse-aws-ipi-deprovision` can also destroy the launched winc nodes?

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/winc/provision/cucushift-installer-rehearse-aws-ipi-ovn-winc-provision-chain.yaml (hunk 11 above)
Comment (jfrancoa) [reply ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/winc/provision/cucushift-installer-rehearse-aws-ipi-ovn-winc-provision-chain.yaml:13]: It is indeed for the test purpose, to prepare some stuff that the test cases expect. However, if not added here, where do you suggest to add them? Without this ref being run all test cases will fail...

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/winc/cucushift-installer-rehearse-aws-ipi-ovn-winc-workflow.yaml (hunk 12 above)
Comment (jfrancoa) [reply ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/winc/cucushift-installer-rehearse-aws-ipi-ovn-winc-workflow.yaml:16]: Mostly for sure, when uninstalling the cluster the Windows nodes will also get cleaned up. For example, in Flexy-destroy there is no WINC specific code to clean up nodes. During destroy process everything gets cleaned up.

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/winc/provision/cucushift-installer-rehearse-aws-ipi-ovn-winc-provision-chain.yaml (hunk 11 above)
Comment (jfrancoa) [reply ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/winc/provision/cucushift-installer-rehearse-aws-ipi-ovn-winc-provision-chain.yaml:13]: At the same time, it does not harm the cluster at all..it's just creating a configmap with data needed during testing as well as some Windows and Linux workloads (everything under the winc-test namespace)

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/winc/cucushift-installer-rehearse-aws-ipi-ovn-winc-workflow.yaml (hunk 12 above)
Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/winc/cucushift-installer-rehearse-aws-ipi-ovn-winc-workflow.yaml:16]: good to know.

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/winc/provision/cucushift-installer-rehearse-aws-ipi-ovn-winc-provision-chain.yaml (hunk 11 above)
Comment (jianlinliu) [reply ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/winc/provision/cucushift-installer-rehearse-aws-ipi-ovn-winc-provision-chain.yaml:13]: got it.

File: ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/winc/cucushift-installer-rehearse-aws-ipi-ovn-winc-workflow.yaml (hunk 12 above)
Comment (jfrancoa) [reply ci-operator/step-registry/cucushift/installer/rehearse/aws/ipi/ovn/winc/cucushift-installer-rehearse-aws-ipi-ovn-winc-workflow.yaml:16]: Windows nodes rely on a machineset too, it's the same as with the Linux workers. They don't need of any extra action to get cleaned up. When uninstalling the cluster the machines get destroyed and with that, the worker nodes. For the cases when we use BYOH nodes, we do it inside the test case and the node is cleaned up in the post-test section.

**Comment by jfrancoa** [comment]
@jianlinliu is there anything else needed to obtain approval for this patch?

**Comment by jfrancoa** [comment]
ping @liangxia @jianlinliu please, could you help me move forward with this patch? We would like to have it merged to be able to deploy a cluster with cluster-bot and debug our Ginkgo testcases further. Debugging from the Prow logs is quite complicated.

File: ci-operator/step-registry/openshift-qe/winc/openshift-qe-winc-ref.yaml [hunk 13]
```diff
@@ -0,0 +1,14 @@
+ref:
```
Comment (liangxia) [inline ci-operator/step-registry/openshift-qe/winc/openshift-qe-winc-ref.yaml:1]: If you move the ref to a folder in `ci-operator/step-registry/cucushift/installer/rehearse/` (most of workflow from QE are located there), then we do not need to wait the approval from openshift-qe organization (mainly used by OpenShift PerfScale team).

Comment (jfrancoa) [reply ci-operator/step-registry/openshift-qe/winc/openshift-qe-winc-ref.yaml:1]: @liangxia The thing is, that this openshift-qe-winc-ref is going to be used in all WINC jobs, that is why I can't simply add it under ci-operator/step-registry/cucushift/installer/rehearse/aws folder. Could I do something similar as hypershift team did? Create a winc-extended folder under [ci-operator/step-registry/cucushift](https://github.com/openshift/release/tree/master/ci-operator/step-registry/cucushift) ? Would there be any problem into that? This way, all the extra refs required for winc team (there will be more) could  be included there.

Comment (liangxia) [reply ci-operator/step-registry/openshift-qe/winc/openshift-qe-winc-ref.yaml:1]: Fine to me to put it into the folder `ci-operator/step-registry/cucushift`. No need to add  `-extended`.

File: ci-operator/step-registry/ipi/conf/aws/windows-machineset/ipi-conf-aws-windows-machineset-ref.yaml [hunk 14]
```diff
@@ -16,6 +16,8 @@ ref:
       default: "m5a.large"
     - name: WINDOWS_NODE_REPLICAS
       default: "1"
+    - name: WINDOWS_OS_ID
+      default: "Windows_Server-2019-English-Full-Containers"
```
Comment (liangxia) [inline ci-operator/step-registry/ipi/conf/aws/windows-machineset/ipi-conf-aws-windows-machineset-ref.yaml:20]: To speed up the PR merge, we can take the changes in `ci-operator/step-registry/ipi/conf/aws/windows-machineset` out and use a follow up PR for it.

Comment (jfrancoa) [reply ci-operator/step-registry/ipi/conf/aws/windows-machineset/ipi-conf-aws-windows-machineset-ref.yaml]: Done, thanks for the suggestion. Submitted a new PR with the other change at https://github.com/openshift/release/pull/37125

**Comment by jfrancoa** [comment]
@liangxia one doubt, I have added the rehearse to 4.14 jobs https://github.com/openshift/release/pull/32453/files#diff-189abf9d5169fda66abe49022c9e2caec613645450e8ac387500d2c1f039a998 ,however I don't see the 4.14 rehearse being triggered...I guess it isn't ready yet, right?

**Comment by liangxia** [comment]
/pj-rehearse pull-ci-openshift-openshift-tests-private-release-4.14-debug-winc-aws-ipi

**Pull request merged**